// spiritconfigs/tests/test.helpers.js
'use strict';

class SpiritConfigsTestHelpers {
    constructor() {
        this.chalk = require('chalk');
        this.files = require('spiritfiles');
        this.fs = require('fs');
        this.rimraf = require('rimraf');
    }


    /**
     * Creates config file for testing.
     */
    createConfig(file, contents) {
        this.files.writeFile(file, contents);
    }

    /**
     * Creates config files for testing.
     */
    createConfigs() {
        this.files.addFiles({
            'configs/a.json': '{"a": "b"}',
            'configs/test1.json': '{"a": "b"}',
            'configs/subdir/test1.json': '{}',
            'customConfigs/test2.json': '{"c": "d"}',
            'node_modules/spiritConfigsTestModule/configs/a.json': '{"a": "123", "b": "c"}',
            'node_modules/spiritConfigsTestModule/configs/test1.json': '{"a": "123", "b": "c"}',
            'node_modules/spiritConfigsTestModule/configs/subdir/test1.json': '{}',
            'node_modules/spiritConfigsTestModule/customConfigs/test2.json': '{}',
            'node_modules/spiritConfigsTestModule/node_modules/configsNestedTestModule/configs/test1.json': '{"b": "d", "c": "d"}',
            'node_modules/spiritConfigsTestModule/node_modules/configsNestedTestModule/configs/subdir/test1.json': '{}',
            'node_modules/spiritConfigsTestModule/node_modules/configsNestedTestModule/customConfigs/test1.json': '{}'
        });
    }

    /**
     * Cleans up the configs created for the testing.
     */
    cleanupConfigs() {
        this.fs.unlinkSync('configs/a.json');
        this.fs.unlinkSync('configs/test1.json');
        this.fs.unlinkSync('configs/subdir/test1.json');
        this.fs.rmdirSync('configs/subdir');
        this.rimraf.sync('customConfigs');
        this.rimraf.sync('node_modules/spiritConfigsTestModule');
    }
}

module.exports = new SpiritConfigsTestHelpers();
