// js3030-installation/ext/minhthai.vscode-todo-parser-1.8.6/out/src/lib/collections/collections.js
"use strict";
// Copyright 2013 Basarat Ali Syed. All Rights Reserved.
//
// Licensed under MIT open source license http://opensource.org/licenses/MIT
//
// Orginal javascript code was by Mauricio Santos
//
var _arrays = require("./src/arrays");
exports.arrays = _arrays;
var Bag_1 = require("./src/Bag");
exports.Bag = Bag_1.default;
var BSTree_1 = require("./src/BSTree");
exports.BSTree = BSTree_1.default;
var Dictionary_1 = require("./src/Dictionary");
exports.Dictionary = Dictionary_1.default;
var Heap_1 = require("./src/Heap");
exports.Heap = Heap_1.default;
var LinkedDictionary_1 = require("./src/LinkedDictionary");
exports.LinkedDictionary = LinkedDictionary_1.default;
var LinkedList_1 = require("./src/LinkedList");
exports.LinkedList = LinkedList_1.default;
var MultiDictionary_1 = require("./src/MultiDictionary");
exports.MultiDictionary = MultiDictionary_1.default;
var Queue_1 = require("./src/Queue");
exports.Queue = Queue_1.default;
var PriorityQueue_1 = require("./src/PriorityQueue");
exports.PriorityQueue = PriorityQueue_1.default;
var Set_1 = require("./src/Set");
exports.Set = Set_1.default;
var Stack_1 = require("./src/Stack");
exports.Stack = Stack_1.default;
var _utils = require("./src/util");
exports.utils = _utils;
